File: backend/libexecution/libbool.ml
open Core_kernel
open Lib
open Types.RuntimeT
module RT = Runtime

let fns : Types.fluid_expr fn list =
  [ { prefix_names = ["Bool::not"]
    ; infix_names = []
    ; parameters = [par "b" TBool]
    ; return_type = TBool
    ; description =
        "Returns the inverse of `b`: true if `b` is false and false if `b` is true"
    ; func =
        InProcess (function _, [DBool b] -> DBool (not b) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Bool::and"]
    ; infix_names = ["&&"]
    ; parameters = [par "a" TBool; par "b" TBool]
    ; return_type = TBool
    ; description = "Returns true if both a and b are true"
    ; func =
        InProcess
          (function
          | _, [DBool a; DBool b] -> DBool (a && b) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Bool::or"]
    ; infix_names = ["||"]
    ; parameters = [par "a" TBool; par "b" TBool]
    ; return_type = TBool
    ; description = "Returns true if either a is true or b is true"
    ; func =
        InProcess
          (function
          | _, [DBool a; DBool b] -> DBool (a || b) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Bool::xor"]
    ; infix_names = []
    ; parameters = [par "a" TBool; par "b" TBool]
    ; return_type = TBool
    ; description =
        "Returns `true` if exactly one of `a` and `b` is `true`. Returns `false` if both are `true` or neither is `true`."
    ; func =
        InProcess
          (function
          | _, [DBool a; DBool b] -> DBool (a <> b) | args -> fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Bool::isNull"]
    ; infix_names = []
    ; parameters = [par "check" TAny]
    ; return_type = TBool
    ; description = "Returns true if the `check` parameter is null"
    ; func =
        InProcess
          (function
          | _, [value] ->
            (match value with DNull -> DBool true | _ -> DBool false)
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = false }
  ; { prefix_names = ["Bool::isError"]
    ; infix_names = []
    ; parameters = [par "check" TAny]
    ; return_type = TBool
    ; description = "Returns `true` if the `check` parameter is an error"
    ; func =
        InProcess
          (function
          | _, [value] ->
            (match value with DError _ -> DBool true | _ -> DBool false)
          | args ->
              fail args)
    ; preview_safety = Safe
    ; deprecated = true } ]

File: fsharp-backend/src/LibExecution/StdLib/LibBool.fs
module LibExecution.LibBool

open System.Threading.Tasks
open FSharp.Control.Tasks
open LibExecution.Runtime
open FSharpPlus
open Prelude

let fn = FnDesc.stdFnDesc

let fns : List<BuiltInFn> =
  [ { name = fn "Bool" "not" 0

      parameters = [ Param.make "b" TBool "" ]
      returnType = TBool
      description =
        "Returns the inverse of `b`: true if `b` is false and false if `b` is true"
      fn =
        (function
        | _, [ DBool b ] -> Value(DBool(not b))
        | args -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = NotDeprecated }
    { name = fn "Bool" "and" 0
      parameters = [ Param.make "a" TBool ""; Param.make "b" TBool "" ]
      returnType = TBool
      description = "Returns true if both a and b are true"
      fn =

        (function
        | _, [ DBool a; DBool b ] -> Value(DBool(a && b))
        | args -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = NotDeprecated }
    { name = fn "Bool" "or" 0
      parameters = [ Param.make "a" TBool ""; Param.make "b" TBool "" ]
      returnType = TBool
      description = "Returns true if either a is true or b is true"
      fn =

        (function
        | _, [ DBool a; DBool b ] -> Value(DBool(a || b))
        | args -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = NotDeprecated }
    { name = fn "Bool" "xor" 0

      parameters = [ Param.make "a" TBool ""; Param.make "b" TBool "" ]
      returnType = TBool
      description =
        "Returns `true` if exactly one of `a` and `b` is `true`. Returns `false` if both are `true` or neither is `true`."
      fn =

        (function
        | _, [ DBool a; DBool b ] -> Value(DBool(a <> b))
        | args -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = NotDeprecated }
    { name = fn "Bool" "isNull" 0

      parameters = [ Param.make "check" TAny "" ]
      returnType = TBool
      description = "Returns true if the `check` parameter is null"
      fn =

        (function
        | _, [ value ] ->
            Value
              (match value with
               | DNull -> DBool true
               | _ -> DBool false)
        | args -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = NotDeprecated }
    { name = fn "Bool" "isError" 0

      parameters = [ Param.make "check" TAny "" ]
      returnType = TBool
      description = "Returns `true` if the `check` parameter is an error"
      fn =

        (function
        | _, [ value ] ->
            Value
              (match value with
               | _ -> DBool true
               | DFakeVal (DError _) -> DBool true)
        | args -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = DeprecatedBecause "an old workaround to poor static typing that's no longer needed" } ]
